src/Databases/MySQL/MaterializeMySQLSyncThread.cpp

```diff
@@ -651,6 +634,36 @@ void MaterializeMySQLSyncThread::onEvent(Buffers & buffers, const BinlogEventPtr
     }
 }
 
+void MaterializeMySQLSyncThread::executeDDLAtomic(Buffers & buffers, const QueryEvent & query_event)
+{
+    buffers.commit(global_context);
```

Comment:
Maybe flush buffers in a separate "transaction" before executing DDL query? Writing GTID sets into `.metadata` when flushing buffers or executing DDL query is a good solution, but it's still not atomic with insertion into MergeTree (or with DDL query), so we should minimize time between creation of new part in MergeTree table and modification of `.metadata` file to make server crash between these points in time less probable.

Reply:
If do that it maybe mean that if the metadata will be modify but the query run failed or server crash the gtid will be recorded.

After ck server restart the sync position is after the gtid, then then it means the data lost. It can make sync stop.

Reply:
If do that it maybe mean that if the metadata will be modify but the query run failed or server crash the gtid will be recorded.

After ck server restart the sync position is after the gtid, then then it means the data lost. It can make sync stop.

Reply:
> if the metadata will be modify but the query run failed or server crash the gtid will be recorded

It should never happen. I suggest the following:

1. Get binlog position **before** `MYSQL_QUERY_EVENT`
2. Do `metadata.transaction(position_before_ddl, [&]() { buffers.commit(global_context); });` to insert new part into MergeTree table and commit all previous `MYSQL_*_ROWS_EVENT`s
3. Get binlog position corresponding to `MYSQL_QUERY_EVENT`
4. Do `metadata.transaction(ddl_position, [&](){ executeDDLAtomic(buffers, query_event); });` to execute DDL query and commit it separately from all previous `MYSQL_*_ROWS_EVENT`s